URL shortener: JSON statistics

rgianassi opened the pull request.

arl commented on go/src/url_shortener/url_shortener.go:
```diff
@@ -114,8 +123,52 @@ func (c *URLShortener) shortenHandler(w http.ResponseWriter, r *http.Request) {
+
+	statsJSON := statsJSON{}
+
+	c.computeStatistics(&statsJSON)
```
In Go it's preferable to to have computeStatistics directly returns a Statistics structure (pointer or not).
Even if it's a pointer, the compiler is clever enough to allocate statsJSON on the stack, since it sees that its lifetime ends with the scope of `statisticsHandler`.

The resulting code would IMHO be easier to use, just call:
```go
stats := c.computeStatistics()
```

arl commented on go/src/url_shortener/url_shortener.go:
```diff
@@ -114,8 +123,52 @@ func (c *URLShortener) shortenHandler(w http.ResponseWriter, r *http.Request) {
 func (c *URLShortener) statisticsHandler(w http.ResponseWriter, r *http.Request) {
-	fmt.Fprintf(w, "%s", c.getStatistics())
+	url := r.URL
+	query := url.Query()
```
You can use query.Get("format') which returns a string here so you don't have to deal with a []string and can remove the loop below

rgianassi replied:
What happens if 2 `format` parameters are passed to the query and the `json` is the second?
Something like: `http://localhost:9090/statistics?format=foo&format=json`.
Or do I need to do not care?

arl replied:
True, with my proposition only the first would be taken into account. But I think we don't care since the specification does't indicate that supplying twice the same parameter is supported, so it's undefined behaviour :-)

arl commented on go/src/url_shortener/url_shortener.go:
```diff
@@ -114,8 +123,52 @@ func (c *URLShortener) shortenHandler(w http.ResponseWriter, r *http.Request) {
 	c.incrementHandlerCounter(c.shortenRoute, true)
```
Instead of having all methods on URLshortener, could we try to organize the code so that:
 - `URLshortener` only has handlers and `getURL/addURL`
 - move the stats-specific methods to `urlStatistics` 

Doing so, we could later add unit-tests of urlStatistics without having to deal with a full URLshortener objects, in order to keep things separate and more easily testable?

arl commented on go/src/url_shortener/url_shortener.go:
```diff
@@ -114,8 +123,52 @@ func (c *URLShortener) shortenHandler(w http.ResponseWriter, r *http.Request) {
 	c.incrementHandlerCounter(c.shortenRoute, true)
 }
 
+type redirectsJSON struct {
+	Success int `json:"success"`
+	Failed  int `json:"failed"`
+}
+
+type handlerJSON struct {
+	Name  string `json:"name"`
+	Count int    `json:"count"`
+}
+
+type serverStatsJSON struct {
+	TotalURL  int           `json:"total_url"`
+	Redirects redirectsJSON `json:"redirects"`
+	Handlers  []handlerJSON `json:"handlers"`
+}
+
+type statsJSON struct {
+	ServerStats serverStatsJSON `json:"server_stats"`
+}
```
At the moment, each time the statistics handler is called we need to recreate the structures in order to serialize to JSON.

Can we try to use the structures highlighted in this comment to keep track of the statistics? and them remove the current fields of  `urlStatistics`.

arl commented on go/src/url_shortener/url_shortener.go:
```diff
@@ -67,38 +68,46 @@ func (c *URLShortener) incrementHandlerCounter(handler string, succeeded bool) {
 
-// handlers
+func (c *URLShortener) computeStatistics(stats *statsJSON) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
```
See other comment about adding methods to the `urlStatistics` struct.

We are locking the whole server _just_ to compute statistics. However if `urlStatistics`  and `URLshortener` were bit more decoupled we wouldn't have to do that, we'd only have to lock the stats struct to update the statistics

arl reviewed (changes requested):
The comments I let requests a different organization of the data structures, the goal is to approach what a typical Go project structure looks like.

Other points:
Generally the file containing `func main` is called main.go since it makes it easier for readers that look at the entry point.

Could we also split main.go into 3 files:
 - main.go: just `func main`
 -server.go contains the HTTP server code
 - stats.go contains statistics related data structures

rgianassi commented:
I split the file in 4 (I added a file for the `shorten` function). I left stats with pointer for now, probably chained calls are better in Go, but I am still unsure.

arl commented on go/src/url_shortener/stats.go:
```diff
@@ -0,0 +1,98 @@
+type StatsJSON struct {
+	ServerStats serverStatsJSON `json:"server_stats"`
+
+	mux sync.Mutex
```
Here a mutex works, but each statistics could also be updated atomically, which would be better performance wise in case our server was under heavy load, when every microsecond counts.

Have a look at the [atomic package](https://godoc.org/sync/atomic)

arl commented on go/src/url_shortener/stats.go:
```diff
@@ -0,0 +1,98 @@
+	stats.TotalURL = totalURL
+}
+
+func (s *StatsJSON) incrementHandlerCounter(handlerURL string, succeeded bool) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	stats := &s.ServerStats
+
+	handlers := &stats.Handlers
+	if handlers == nil {
+		*handlers = make([]handlerJSON, 0)
+	}
+
+	found := false
+	for i := range *handlers {
+		handler := &(*handlers)[i]
+		name := handler.Name
+		if name != handlerURL {
+			continue
+		}
+
+		handler.Count++
+		found = true
+		break
+	}
+
+	if !found {
+		*handlers = append(*handlers, handlerJSON{handlerURL, 1})
+	}
+
+	redirects := &stats.Redirects
+	if succeeded {
+		redirects.Success++
+	} else {
+		redirects.Failed++
+	}
+}
```
I think this function could be rewritten in a much simpler way:
 - constructors are not part of the language, but as soon as a `struct` has a non-trivial initialization the convention is to write what we call a _constructor function_. It simply is a function of the form: `newMyStruct(...) *MyStruct`  or  `newMyStruct(...) MyStruct`  depending on the struct semantics. If such a function existed for `StatsJSON`, `stats.Handlers` could be allocated there. Also, since in our case the total number of handlers is known beforehand, all the handlers could be allocated in the constructor.
 - again, since the handlers are known beforehand, we could assign an index to each handler (using the `const` and `iota` keywords) and pass that index to `incrementHandlerCounter` rather than the string. Doing so would also make it more easy to update the number of succeeded/failed redirects atomically

rgianassi replied:
I made this, but it seems to me that stats and server are coupled again.

arl commented on go/src/url_shortener/stats.go:
```diff
@@ -0,0 +1,98 @@
+
+	stats := s.ServerStats
+
+	pairsInCache := fmt.Sprintf("Number of long/short URL pairs: %v", stats.TotalURL)
+
+	redirects := stats.Redirects
+	succeededRedirects := fmt.Sprintf("Succeeded redirects: %v", redirects.Success)
+	failedRedirects := fmt.Sprintf("Failed redirects: %v", redirects.Failed)
+
+	handlers := &stats.Handlers
+	handlerCalls := make([]string, 0, len(*handlers))
+
+	for _, handler := range *handlers {
+		handlerCalls = append(handlerCalls, fmt.Sprintf("Handler %s called %v time(s)", handler.Name, handler.Count))
+	}
+
+	statsBody := fmt.Sprintf("Some statistics:\n\n%s\n%s\n%s\n%s", pairsInCache, succeededRedirects, failedRedirects, strings.Join(handlerCalls, "\n"))
+	return statsBody
```
For this kind of case it's easier, but also more performance in terms of volume of allocations, to use a [strings.Builder](https://godoc.org/strings#Builder) example: https://play.golang.org/p/LnoC147qON5

arl reviewed (approved):
Really clean and nice code :+1: 

The comments I let are either performance-oriented or some advice about stuff in the standard library you're not yet aware of.
